### PR TITLE
Datastore save events

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
related #1113
save events for datastore

### CODE REVIEW

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 1]
```diff
@@ -140,32 +142,35 @@ public void setApplicationEventPublisher(ApplicationEventPublisher applicationEv
 
 	@Override
 	public <T> T save(T instance, Key... ancestors) {
-		return save(instance, new HashSet<>(), ancestors);
-	}
-
-	private <T> T save(T instance, Set<Key> persisted, Key... ancestors) {
-		saveAll(Collections.singletonList(instance), persisted, ancestors);
+		List<T> instances = Collections.singletonList(instance);
```
Comment (dmitry-s) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: this variable seems unnecessary

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 2]
```diff
@@ -382,23 +387,26 @@ public Key createKey(String kind, Object id) {
...
-		saveReferences(entity, builder, persistedEntities);
-		return builder.build();
+		entitiesToSave.addAll(saveDescendents(entity, key, persistedEntities));
+		entitiesToSave.addAll(saveReferences(entity, builder, persistedEntities));
+		entitiesToSave.add(builder.build());
+		return entitiesToSave;
 	}
 
-	private void saveReferences(Object entity, Builder builder, Set<Key> persistedEntities) {
+	private List<Entity> saveReferences(Object entity, Builder builder, Set<Key> persistedEntities) {
```
Comment (dmitry-s) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: this method doesn't save references anymore, so the name is misleading

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 3]
```diff
@@ -409,24 +417,26 @@ private void saveReferences(Object entity, Builder builder, Set<Key> persistedEn
...
 				Key key = getKey(val, false);
 				value = KeyValue.of(key);
 			}
 			builder.set(((DatastorePersistentProperty) persistentProperty).getFieldName(), value);
 		});
+		return entitiesToSave;
 	}
 
-	private void saveDescendents(Object entity, Key key, Set<Key> persistedEntities) {
+	private List<Entity> saveDescendents(Object entity, Key key, Set<Key> persistedEntities) {
```
Comment (dmitry-s) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: same here

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 4]
```diff
@@ -140,32 +142,35 @@ public void setApplicationEventPublisher(ApplicationEventPublisher applicationEv
...
 
 	@Override
 	public <T> Iterable<T> saveAll(Iterable<T> entities, Key... ancestors) {
-		return saveAll(entities, new HashSet<>(), ancestors);
+		saveEntities((List<T>) entities, saveAll(entities, new HashSet<>(), ancestors));
+		return entities;
 	}
 
-	private <T> Iterable<T> saveAll(Iterable<T> entities, Set<Key> persisted, Key... ancestors) {
+	private <T> List<Entity> saveAll(Iterable<T> entities, Set<Key> persisted, Key... ancestors) {
```
Comment (dmitry-s) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: this method doesn't save, the name is misleading

File: spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java [hunk 5]
```diff
@@ -504,6 +504,7 @@ public void inheritanceTest() {
 
 	@Test
 	public void inheritanceTestFindAll() {
+		this.datastoreTemplate.deleteAll(Pet.class);
```
Comment (dmitry-s) [inline spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java]: we clean this table in `@After` method

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 1 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: It's used twice in the next line so I gave it its own name

File: spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java (hunk 5 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java]: done, PTAL

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java]: @dmitry-s PTAL

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 1 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: I see, makes sense

**Pull request merged**